### PR TITLE
Modules

### PR DESCRIPTION
**Pull request by thetrime**
This pull request implements modules in the style of SWI-Prolog. 

Code should be backward compatible with any existing code, (provided that existing code does not declare :-module/2 anywhere, or define a predicate :/2 obviously!)

The actual mechanics are probably quite different. When a predicate is called that we do not know about in the current module, before we return getUndefinedPredicateCode(), check the user module to see if it's declared there. if it is, return that. 
When we compile a module, all the exported predicates are turned into shims in user of the form 
Head:- :(ModuleName, Head).

:/2 is defined to push a new module context onto a stack. This is popped once the goal exits (and is restored appropriately during backtracking). 

I'm more than happy to discuss the design, implementation, test cases and whether this is documented well enough - this is really a starting point for a patch rather than a final submission!

### CODE REVIEW

File: src/gnu/prolog/database/Module.java [hunk 1]
```diff
@@ -44,6 +55,86 @@
...
+		try
+		{
+			for (CompoundTermTag export: exports)
+			{
+				importPredicate(environment, exportingModule, export);
+			}
+		}
+		catch(Throwable builtin)
+		{
+			// FIXME: It is quite bad if this happens
```
Comment (drt24) [inline src/gnu/prolog/database/Module.java]: Indeed. We can't just ignore this.

Comment (thetrime) [inline src/gnu/prolog/database/Module.java]: Ah, quite. Digging deeper into this uncovered some other problems that I meant to get to the bottom of but never did. I think it's much clearer now, and with the added bonus of that exception not disappearing into the ether. Have a look at the revised version with my extra commit.

File: src/gnu/prolog/database/Module.java [hunk 2]
```diff
@@ -183,4 +262,150 @@ public void removePredicateListener(PredicateListener listener)
...
+		if (code == null)
+		{
+			code = env.loadPrologCode(tag);
+			tag2code.put(tag, code);
+		}
+		return code;
+
+	}
+
+	protected void pollPrologCodeListeners()
```
Comment (drt24) [inline src/gnu/prolog/database/Module.java:398]: This method wants a comment explaining what it is for. I think it is checking the list to update it to remove references that no longer exist?

File: src/gnu/prolog/database/Module.java [hunk 3]
```diff
@@ -20,15 +20,26 @@
```
Comment (drt24) [inline src/gnu/prolog/database/Module.java:21]: You need to update the copyright info on this file, major changes.

File: src/gnu/prolog/database/PrologTextLoader.java [hunk 4]
```diff
@@ -328,7 +328,15 @@ protected void processModuleDirective(Term moduleName, Term exports)
...
 		}
-		prologTextLoaderState.startNewModule(this, (AtomTerm)moduleName, exports);
+		try
+		{
+			prologTextLoaderState.startNewModule(this, (AtomTerm)moduleName, exports);
+		}
+		catch (PrologException e)
+		{
+			e.printStackTrace();
+			logError("unable to start new module");
```
Comment (drt24) [inline src/gnu/prolog/database/PrologTextLoader.java:338]: I see what you mean about this use of logError rather than throwing exceptions. I can't quite remember why we do it this way? Is it so that all the compile errors get found at once rather than one at a time?

File: src/gnu/prolog/vm/buildins/meta/Predicate_colon.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
+/* GNU Prolog for Java
+ * Copyright (C) 2011 Matt Lilley
```
Comment (drt24) [inline src/gnu/prolog/vm/buildins/meta/Predicate_colon.java]: 2016?

File: src/gnu/prolog/database/Module.java (hunk 2 above)
Comment (drt24) [inline src/gnu/prolog/database/Module.java:398]: I now see it came from another file but I still think it wants a comment ;-)

**Comment by drt24** [comment]
This looks good. I have only made a couple of minor comments. I think this feature wants a brief mention in the manual: docs/manual.texinfo but that can be done separately. I am happy to merge this now or you can address the minor questions first.

File: src/gnu/prolog/database/Module.java (hunk 2 above)
Comment (thetrime) [inline src/gnu/prolog/database/Module.java:398]: Yep. I think you added it with 1fb588a72b1176d0b0f42741d494294eb211de59. I'll add some sort of comment though

File: src/gnu/prolog/database/PrologTextLoader.java (hunk 4 above)
Comment (thetrime) [inline src/gnu/prolog/database/PrologTextLoader.java:338]: It could be so that you could 'catch' those errors in Java. I'm not really sure either - it was well before I had any involvement. I think that, generally speaking, fixing that is an orthogonal problem to implementing modules, and if you want to look at converting those logged messages into exceptions, we should file a new issue and do it in a new branch

File: src/gnu/prolog/vm/buildins/meta/Predicate_colon.java (hunk 5 above)
Comment (thetrime) [inline src/gnu/prolog/vm/buildins/meta/Predicate_colon.java]: Oops. Yes - I copied this from the header of another file.

**Comment by thetrime** [comment]
Ok. I've hopefully tidied up the issues reported. Let me know what you think

**Comment by drt24** [comment]
LGTM.
